### ISSUE TITLE
Memory leak in ServiceRemotingPartitionClient

### ISSUE THREAD

**Issue by jamesweb-ms** [Azure/service-fabric-issues#344]
This looks very similar to issue #62, but I'm running 2.4.164 and the issue was closed in 2.3.

in a for loop I'm calling:
                            var actor = ActorProxy.Create<IMonitorMdsTable>(new ActorId(string.Format("{0}_{1}", this.events[eventType].Name, tableName)), this.Context.CodePackageActivationContext.ApplicationName);
                            monitors.Add(actor.StartMonitorTableAsync(new EventSettingWrapper() { EventSetting = this.events[eventType] }, cancellationToken));

After running for 24hours, I see that my memory has steadily increased 3gb.  Looking at memory dump, I have 1.2 million instances of ServiceRemotingPartitionClient.  I only have ~10k actors.  Before I go and implement a cache for my proxyClients, I just wanted to see if maybe this was really fixed in a later version.

![image](https://user-images.githubusercontent.com/10124358/27977425-904d805e-631f-11e7-8cef-c0a6a10b8f5b.png)

**Comment by MedAnd** [comment]
By any chance are you using Server GC as per this issue: https://github.com/Azure/service-fabric-issues/issues/320

**Comment by jamesweb-ms** [comment]
I assume this would be set in <application>.exe.config?  If so, no I do not have server gc set.

Also I tried work around of caching connections and still see leak.

**Comment by MedAnd** [comment]
We've had to create an App.config file per microservice and add the following:

`<configuration>
   <runtime>
      <gcServer enabled="false"/>
   </runtime>
</configuration>
`

**Comment by suchiagicha** [comment]
@jamesweb-ms

Are you sharing cancellation token /single cancellationtokensource between different request?
Could you try if you create new cancellation token for every request?

**Comment by jamesweb-ms** [comment]
I am...
```
	foreach (string mdsTable in tables)
	{
		var actor = ActorProxy.Create<IMonitorMdsTable>(new ActorId(string.Format("{0}_{1}", this.events[eventType].Name, mdsTable)), this.Context.CodePackageActivationContext.ApplicationName);
		monitors.Add(actor.StartMonitorTableAsync(new EventSettingWrapper() { EventSetting = this.events[eventType], StorageConnectionString = this.storageConnectionString }, cancellationToken));
	}
```
So you are suggesting...
```
	foreach (string mdsTable in tables)
	{
		CancellationTokenSource cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
		var actor = ActorProxy.Create<IMonitorMdsTable>(new ActorId(string.Format("{0}_{1}", this.events[eventType].Name, mdsTable)), this.Context.CodePackageActivationContext.ApplicationName);
		monitors.Add(actor.StartMonitorTableAsync(new EventSettingWrapper() { EventSetting = this.events[eventType], StorageConnectionString = this.storageConnectionString }, cts.Token));
	}

```

**Comment by suchiagicha** [comment]
@jamesweb-ms 
We had made a fix for the bug in latest sdk 2.7 (which is yet to be released soon)  for this issue.
Issue is we were not unregistering the cancellationtoken registration in our code, so when the cancellation token is long lived , the objects are not cleaned up .
No. For the workaround, I was actually asking you to create new cancellation token source for all request. I don't think linked cancellation token will help you. Since internally they are sharing cancellation token.

**Comment by jamesweb-ms** [comment]
Ugly... so this
```
List<CancellationTokenSource> monitorCTS = new List<CancellationTokenSource>();
List<Task> monitors = new List<Task>();
try
{
	foreach (string mdsTable in tables)
	{
		CancellationTokenSource cts = new CancellationTokenSource();
		monitorCTS.Add(cts);
		var actor = ActorProxy.Create<IMonitorMdsTable>(new ActorId(string.Format("{0}_{1}", this.events[eventType].Name, mdsTable)), this.Context.CodePackageActivationContext.ApplicationName);
		monitors.Add(actor.StartMonitorTableAsync(new EventSettingWrapper() { EventSetting = this.events[eventType], StorageConnectionString = this.storageConnectionString }, cts.Token));
	}
	await Task.WhenAny(Task.WhenAll(monitors), Task.Delay(TimeSpan.FromMinutes(10), cancellationToken));
}
catch (OperationCanceledException)
{
	foreach (var cts in monitorCTS)
	{
		cts.Cancel();
	}
}
```

I've deployed this change and will let you know tomorrow...

**Comment by jamesweb-ms** [comment]
Still appear to have leak after running previous change overnight.

**Comment by suchiagicha** [comment]
We wrote test where we were not storing cancellationtokensource ,so all the objects were holds up till the time CancellationTokenSource  object was in memory and after that cleanup was happening. May be you can check the lifetime of object monitorCTS  in your process ?

**Comment by jamesweb-ms** [comment]
After leaving it running for several days, it looks like above code did solve issue...  overnight was an aberration.  Using Linked Tokens did not work as you mentioned.

**Comment by amolenk** [comment]
Just ran into the same problem. Removing our cancellation tokens from the Actor calls as mentioned above fixed it.

**Comment by suchiagicha** [comment]
This has been fixed in 2.7 sdk .
